addons/token-exchange/agent_mirrorpeer_controller.go

```diff
@@ -84,13 +88,71 @@ func (r *MirrorPeerReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 		return ctrl.Result{}, fmt.Errorf("failed to enable mirroring the storagecluster %q in namespace %q in managed cluster. Error %v", scr.Name, scr.Namespace, err)
 	}
 
+	err = r.createS3(ctx, req)
+	if err != nil {
+		klog.Error(err, "Failed to create ODR S3 resources")
+		return ctrl.Result{}, err
+	}
+
 	err = r.createVolumeReplicationClass(ctx, &mirrorPeer)
 	if err != nil {
 		return ctrl.Result{}, fmt.Errorf("failed to create volumereplicationclass: %v", err)
 	}
 	return ctrl.Result{}, nil
 }
 
+func (r *MirrorPeerReconciler) createS3(ctx context.Context, req ctrl.Request) error {
+	var err error
+
+	var mirrorPeer multiclusterv1alpha1.MirrorPeer
+	err = r.HubClient.Get(ctx, req.NamespacedName, &mirrorPeer)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			klog.Info("Could not find MirrorPeer")
+			return err
+		}
+		klog.Error(err, "Failed to get MirrorPeer")
+		return err
+	}
+	var peerAccumulator string
+	for _, peer := range mirrorPeer.Spec.Items {
+		peerAccumulator += peer.ClusterName
+	}
+	checksum := sha1.Sum([]byte(peerAccumulator))
+
+	bucketGenerateName := "odrsystem-"
+	// truncate to bucketGenerateName + first 12 (out of 20) byte representations of sha1 checksum
+	bucket := fmt.Sprintf("%s%s", bucketGenerateName, hex.EncodeToString(checksum[:]))[0 : len(bucketGenerateName)+12]
+
+	namespace := common.GetEnv("ODR_NAMESPACE", common.RamenHubNamespace)
+
+	noobaaOBC := &obv1alpha1.ObjectBucketClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      bucket,
+			Namespace: namespace,
+		},
+		Spec: obv1alpha1.ObjectBucketClaimSpec{
+			BucketName:       bucket,
+			StorageClassName: "openshift-storage.noobaa.io",
```

Comment:
Depends on StorageCluster namespace available in MirrorPeer. It might not always be `openshift-storage`.

Reply:
```go
var scNamespace string
m := make(map[string]bool)
for _, item := range mirrorPeer.Spec.Items {
	ns := item.StorageClusterRef.Namespace
	if _, ok := m[ns]; ok {
		continue
	}
	m[ns] = true
	if err = r.SpokeClient.Get(ctx, types.NamespacedName{Name: "ocs-storagecluster", Namespace: ns}, &ocsv1.StorageCluster{}); err == nil {
		scNamespace = ns
		break
	}
}
// Refer https://github.com/red-hat-storage/odf-multicluster-orchestrator/pull/59#issuecomment-1012165500
namespace := common.GetEnv("ODR_NAMESPACE", scNamespace)
noobaaOBC := &obv1alpha1.ObjectBucketClaim{
	ObjectMeta: metav1.ObjectMeta{
		Name: bucket,
		Namespace: namespace,
	},
	Spec: obv1alpha1.ObjectBucketClaimSpec{
		BucketName:       bucket,
		StorageClassName: namespace + ".noobaa.io",
	},
}
```

Is there a resource with an already known and fixed name and namespace that specifies the namespace where the SC is present? I hacked something (as you can see above), but I believe there might be a better way of doing this.

Reply:
There's a configmap that holds the clustername that I can use to point to the SC namespace from the MP spec. I can see this in my minikube cluster but not on the openshift one, maybe it's a different configmap. Looking.